### PR TITLE
Avoid re-reading parquet footer for iceberg stats in writer

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Description
Currently we're reading parquet file footer back just after
writing it to file system for computing iceberg stats.
Since Trino parquet writer already has the file footer, there
is no need to perform an extra IO to get the footer again.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Improve performance of writing parquet files. ({issue}`19090`)
```

### CODE REVIEW

**Review by sopel39 — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java [hunk 1]
```diff
@@ -70,14 +72,19 @@ public IcebergParquetFileWriter(
...
                 Optional.empty());
+        this.location = outputFile.location();
         this.metricsConfig = requireNonNull(metricsConfig, "metricsConfig is null");
-        this.inputFile = new ForwardingInputFile(fileSystem.newInputFile(outputFile.location()));
     }
 
     @Override
     public Metrics getMetrics()
     {
-        return fileMetrics(inputFile, metricsConfig);
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java:78]: TestIcebergMetadataFileOperations seems to be specifically about metadata operations, this is a "DATA" operation.
There is code preventing assertions on DATA operations at https://github.com/trinodb/trino/blob/fe70595917a8d2b5ce6feecb9f4e10036c8bd86a/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetadataFileOperations.java#L633

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java [hunk 2]
```diff
@@ -77,4 +79,40 @@ public Metrics getMetrics()
...
+    @Override
+    public void rollback()
+    {
+        parquetFileWriter.rollback();
+    }
+
+    @Override
+    public long getValidationCpuNanos()
+    {
+        return parquetFileWriter.getValidationCpuNanos();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java:126]: should `getVerificationTask` be forwarded as well?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java:78]: you're right. we can either broaden the test class coverage (rename class to match data operations too)
or add similar class, a copy
or abstain from testing

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java (hunk 2 above)
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java:126]: `getVerificationTask` is unused so I just removed it in a prep commit

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java [hunk 3]
```diff
@@ -34,12 +35,12 @@
 import static java.util.Objects.requireNonNull;
 import static org.apache.iceberg.parquet.ParquetUtil.fileMetrics;
 
-public class IcebergParquetFileWriter
-        extends ParquetFileWriter
+public final class IcebergParquetFileWriter
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java:42]: > Avoid using inheritance to implement IcebergParquetFileWriter

maybe put some rationale in the commit body

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java [hunk 4]
```diff
@@ -70,14 +73,21 @@ public IcebergParquetFileWriter(
...
     @Override
     public Metrics getMetrics()
     {
-        return fileMetrics(inputFile, metricsConfig);
+        ParquetMetadata parquetMetadata;
+        try {
+            parquetMetadata = new ParquetMetadataConverter().fromParquetMetadata(parquetFileWriter.getFileMetadata());
+        }
+        catch (IOException e) {
+            throw new TrinoException(ICEBERG_WRITER_OPEN_ERROR, format("Error creating Parquet file %s", location), e);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java:88]: ICEBERG_WRITER_OPEN_ERROR? "Error creating Parquet file"?
maybe the messge & code should be different here

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java [hunk 5]
```diff
@@ -395,4 +392,45 @@ private void initColumnWriters()
...
+        private FileFooter(MessageType messageType, String createdBy, Optional<DateTimeZone> parquetTimeZone)
+        {
+            this.messageType = messageType;
+            this.createdBy = createdBy;
+            this.parquetTimeZone = parquetTimeZone;
+        }
+
+        public void addRowGroup(RowGroup rowGroup)
+        {
+            requireNonNull(rowGroupBuilder, "rowGroupBuilder is null");
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java:414]: `checkState` would be more appropriate

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java [hunk 6]
```diff
@@ -395,4 +392,45 @@ private void initColumnWriters()
...
+            rowGroupBuilder.add(rowGroup);
+        }
+
+        public FileMetaData createFileMetadata()
+        {
+            requireNonNull(rowGroupBuilder, "rowGroupBuilder is null");
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java:420]: `checkState` would be more appropriate

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java [hunk 7]
```diff
@@ -342,31 +350,20 @@ private void writeFooter()
...
-    {
-        FileMetaData fileMetaData = new FileMetaData();
-        fileMetaData.setVersion(1);
-        fileMetaData.setCreated_by(createdBy);
-        fileMetaData.setSchema(MessageTypeConverter.toParquetSchema(messageType));
-        // Added based on org.apache.hadoop.hive.ql.io.parquet.write.DataWritableWriteSupport
-        parquetTimeZone.ifPresent(dateTimeZone -> fileMetaData.setKey_value_metadata(
-                ImmutableList.of(new KeyValue("writer.time.zone").setValue(dateTimeZone.getID()))));
-        long totalRows = rowGroups.stream().mapToLong(RowGroup::getNum_rows).sum();
-        fileMetaData.setNum_rows(totalRows);
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java:356]: where `setNum_rows` went to?

the new code doesn't apparently call this method?

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java:356]: It became a constructor argument to FileMetadata

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java (hunk 3 above)
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java:43]: added commit message

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java (hunk 4 above)
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java]: updated this

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergParquetFileWriter.java:78]: I'm skipping testing access count explicitly for now

**Pull request merged**